csrc/dd_arith.c

```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
     c = sqrtq(adddq(1.0, sqrq(s)));
     return divqq(s, c);
 }
+
+ddouble tanq(ddouble a)
+{
+    if (iszeroq(a))
+        return Q_ZERO;
+
+    ddouble s, c;
+    s = sinq(a);
+    c = cosq(a);
+    return divqq(s, c);
+}
+
+void sincos(const ddouble a, ddouble *sin_a, ddouble *cos_a) {
+
+    if (iszeroq(a)) {
+        *sin_a = Q_ZERO;
+        *cos_a = Q_ONE;
+        return;
+    }
+
+    // approximately reduce modulo 2*pi
+    ddouble z = nintq(divqq(a, Q_2PI));
+    ddouble r = subqq(a, mulqq(Q_2PI, z));
+
+    // approximately reduce module pi/2 and pi/16
+    ddouble t;
+    double q = floor(r.hi / Q_PI_2.hi + 0.5);
+    t = subqq(r, mulqd(Q_PI_2, q));
+    int j = (int)(q);
+    int abs_j = (int)(fabs(j));
+    q = floor(t.hi / Q_PI_16.hi + 0.5);
+    t = subqq(t, mulqd(Q_PI_16, q));
+    int k = (int)(q);
+    int abs_k = (int)(fabs(k));
+
+    if (abs_j > 2) {
+        // dd_real::error("(dd_real::sincos): Cannot reduce modulo pi/2.");
+        *cos_a = *sin_a = nanq();
+        return;
+    }
+
+    if (abs_k > 4) {
+        // dd_real::error("(dd_real::sincos): Cannot reduce modulo pi/16.");
+        *cos_a = *sin_a = nanq();
+        return;
+    }
+
+    ddouble sin_t, cos_t;
+    ddouble s, c;
+
+    sincos_taylor(t, &sin_t, &cos_t);
+
+    if (abs_k == 0) {
+        s = sin_t;
+        c = cos_t;
+    } else {
+        ddouble u = _cos_table[abs_k - 1];
+        ddouble v = _sin_table[abs_k - 1];
+
+        if (k > 0) {
+            s = addqq(mulqq(u, sin_t), mulqq(v, cos_t));
+            c = subqq(mulqq(u, cos_t), mulqq(v, sin_t));
+        } else {
+            s = subqq(mulqq(u, sin_t), mulqq(v, cos_t));
+            c = addqq(mulqq(u, cos_t), mulqq(v, sin_t));
+        }
+    }
+    if (abs_j == 0) {
+        *sin_a = s;
+        *cos_a = c;
+    } else if (j == 1) {
+        *sin_a = c;
+        *cos_a = negq(s);
+    } else if (j == -1) {
+        *sin_a = negq(c);
+        *cos_a = s;
+    } else {
+        *sin_a = negq(s);
+        *cos_a = negq(c);
+    }
+
+}
+
+
+ddouble atan2qq(ddouble y, ddouble x) {
+    /* Strategy: Instead of using Taylor series to compute
+        arctan, we instead use Newton's iteration to solve
+        the equation
+
+            sin(z) = y/r    or    cos(z) = x/r
+
+        where r = sqrt(x^2 + y^2).
+        The iteration is given by
+
+            z' = z + (y - sin(z)) / cos(z)          (for equation 1)
+            z' = z - (x - cos(z)) / sin(z)          (for equation 2)
+
+        Here, x and y are normalized so that x^2 + y^2 = 1.
+        If |x| > |y|, then first iteration is used since the
+        denominator is larger.  Otherwise, the second is used.
+    */
+
+    if (iszeroq(x)) {
+
+        if (iszeroq(y)) {
+            /* Both x and y is zero. */
+            return Q_ZERO;
+    }
+
```

Comment:
This looks a little messy
```suggestion
    if (iszeroq(x)) {
        if (iszeroq(y))
            return Q_ZERO;           /* Both x and y are zero */
```

Reply:
Fixed